### PR TITLE
Dumpster and vehicle storage

### LINKED ISSUE #69159: Vehicle beds should be capable of holding a human corpse
### Describe the bug

Human corpses take 62.5L, vehicle beds can hold a maximum of 50L

### Attach save file

N/A

### Steps to reproduce

1. Acquire a vehicle that has a bed
2. Attempt to place a corpse into the bed
3. Wonder how people normally sleep

### Expected behavior

You can fit a human into a bed, dead or not

### Screenshots

_No response_

### Versions and configuration

- OS: Linux
    - OS Version: Distributor ID: Ubuntu; Description: Ubuntu 20.04.6 LTS; Release: 20.04; Codename: focal; 
- Game Version: e497f5a [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Slowdown Fungal Growth [no_fungal_growth],
    Bionic Slots [cbm_slots],
    Tamable Wildlife [Tamable_Wildlife],
    Portal Storms Ignore NPCs [personal_portal_storms],
    Disable NPC Needs [no_npc_food],
    Stats Through Kills [stats_through_kills],
    Stats Through Skills [StatsThroughSkills]
]

### Additional context

This is mainly due to the ambulance; shoving a friendly's corpse into the boxes next to the bed is not the most considerate method of transporting a human.

### LINKED ISSUE #70180: Dumpster volume is too low and should be incrased
### Is your feature request related to a problem? Please describe.

I like to use the now constructable dumpster as a place to dump rotten foods unneeded containers and damaged components I don't want available for vehicle constructions.

The in-game dumpster has a volume of 1000L. This is similar to the bare floor.

I assume the in-game dumpsters to be the Dempster style dumpster common in most cities and commercial areas for trash removal as opposed to roll-off style dumpsters one might see at construction sites.

https://en.wikipedia.org/wiki/George_Roby_Dempster

image

The standard city dumpster I see on a daily basis is probably either a 6 yard (4500L) or 8 yard (6000L) dumpster

### Solution you would like.

Increase the volume a dumpster can hold from 1000L to either:

2 yard ~ 1500L
4 yard ~ 3000L
6 yard ~ 4500L
8 yard ~ 6000L

Preferably either 6 yard or 8 yard since a dumpster should be able to hold more than most other game containers. You could argue that actual warehouse shelving, being floor to ceiling, could hold more.

### Describe alternatives you have considered.

N/A

### Additional context

image

### PR DESCRIPTION
**Pull request by worm-girl**
#### Summary
Bugfixes "Vehicle parts check creature size, increases storage of dumpster and vehicle parts"

#### Purpose of change

Closes #69159 
Closes #70180 

Several vehicle parts which should logically be able to store a human corpse, namely beds and seats, were unable to do so. Additionally, dumpsters were not providing any storage advantage despite that being their explicit purpose IRL.

As pointed out in the comments, simply turning up the storage volume on seats was problematic as you could then have a seat store both a corpse and a person. Therefore I took the opportunity to make characters and monsters dynamically check cargo space when moving through vehicles. This makes being a mutant, especially a large or huge one, more impactful, and will similarly affect power armor and CBMs if these are ever able to change a character's effective size.

#### Describe the solution

image

- Went with the smallest bin here for the dumpster, as it only takes up one tile, though most tilesets mush multiples into a single bigger dumpster. So it's 1500 ml per tile, either way
- Added LIQUIDCONT to dumpsters, as they will hold fluids IRL if you want to use them that way, and also if you don't

VEHICLE SEAT STORAGE AND COMFORT
- Made vehicle seats able to hold 80 liters. Reclining seats naturally can make more space, so they get bumped up to 84. The wooden seat and the prison seat are really only just big enough to hold a body, so they get 62.5
- Vehicle beds hold 100 liters
- Makeshift seats, cloth seats, bike saddles etc. still hold nothing
- Also noticed that vehicle back seats were comfort 4, which is the same as a vehicle bed. This made vehicle beds pointless and was causing some players to always replace their front seats with back seats, which doesn't seem like desirable behavior. Reclining and bench seats still have the BED flag, but remain slightly inferior to an actual purpose-build bed
- Renamed back seats to bench seats and gave them their own item. Many vehicles, particularly older ones, use bench seats up front. Also renamed regular seats to bucket seats to make the delineation clear
- Reclining seats and bench seats are now both comfort 3, the same as an armchair. If you're a player, that means reclining seats are ideal for maximizing storage and comfort, but bench seats are still slightly superior to normal seats
- Livestock stalls are now comfort 2, the same as a straw bed. They also have the BED flag
- Converted bench and bucket seats to use nylon rather than cotton in their construction and repair
- Gave aisles 100 liters of storage. The vehicle part is described as one square meter in size. A flat one square meter area can actually hold 333 liters of soil (random example item) if you stacked it in a cone shape one meter high, assuming a ~30 degree angle of repose. But this is a vehicle and there's bound to be other stuff in the way, and it's unsecured, etc., so we'll err on the side of caution here. For reference, a standard trunk tile is 300 liters, so aisles are not replacing proper storage any time soon
- Upgraded floor trunks to 100 liters of storage for parity with aisles. I was not able to find any kind of commercially available floor trunk when searching online. I did see a few custom jobs which I assume is what's going on here, but they seemed quite small. Ideally IMO this vehicle part would be better as a "pocket" for securing items beneath a regular aisle and adding to its storage space, but that isn't currently possible. As-is, this is still a straight upgrade over the aisle as it secures the contents
- Gave non-upholstered bench-style seats 1 comfort, to match what benches have outside of a vehicle

STORAGE VS CREATURE SIZE
- Unless a vehicle has CARGO_PASSABLE (doors, floor trunks), if it is able to contain items, the size of characters and monsters will be checked when they try to enter it. CARGO_PASSABLE vehicle parts get a string explaining this feature from the vpart flag
- This feature currently checks the average volume of a creature of your size, and does not account for BMI or worn equipment. These would be fine expansions to this system, but are out of scope for now
- If 100% of your body volume is free, you can enter just fine
- If there is less than 100% but more than 75% of your body volume in free space, you can squeeze in. This applies a negative effect unless you have BENDY1, BENDY2, or BENDY3. Elves and slimes rejoice
- The cramped space effect slows you, debuffs dex and str, and causes pain. It goes away when you free yourself. It's impossible to sleep in cramped spaces and NPCs will periodically complain about them if stuck there
- If there is less than 75% of your body volume available, you will be unable to enter
- If the space has no roof, vehicle tiles have 20% extra space to fit people. Actual cargo space is unchanged
- Dropping stuff in a space you currently occupy will cause that space to become cramped for you until either you move to a better tile or you move the items which are in the space
- If you are huge, you will always get the cramped space debuff unless you're in an aisle or a tile with the HUGE_OK flag. Currently that includes the livestock stall, yokes, and cargo spaces, so you're good to ride in the trunk or sleep in the cow carrier. Driving is possible but currently it will always be uncomfortable. However, the new effect is much less janky than the old one, which used to just infinitely apply pain
- Updated dialog snippets involving NPCs complaining about cramped space. Since people of any size can be cramped, it was inappropriate to have ones directly referencing their hugeness. I took them out and added some more which are less specific
- Monsters which are more than 850 liters simply cannot fit in any tile which doesn't have HUGE_OK. Very few monsters are actually that big, mostly just hulks, trapped tendrils, the yrax sphenocorona, armored centipede, titan stag beetle and a few others

#### Describe alternatives you've considered

None

#### Testing

- Spawned in, spawned a car, used debug hamerspace to install and uninstall the new seats, saw that they produced the correct items when uninstalled
- Checked storage capacity on all modified items
- Checked crafting recipes to ensure nothing weird happened
- Set up different configurations of vehicle tiles with different amounts of stuff in them. Mutated to different sizes and tested that I could pass only where appropriate for my size
- Stuck a dead person in a car seat with all the gear they would have been able to wear in it while alive
- Tried getting in cars with NPCs when there was stuff on the seats. They successfully pathfinded to other seats as well as they ever have
- Made friendly NPCs of different sizes and observed that they knew which seats would hold them
- Spawned a vehicle, sat in the driver's seat with the passenger door open, placed a corpse in the passenger seat. Zombies outside were unable to move through the passenger seat, but were happy to smash the car until they could get through.
- Did the above test with a grappler zombie, saw that it did not use RANGED_PULL as long as the filled seat was in the way, as it's an impassable tile
- Put a pet cow in a livestock pen, saw that a pet cow was unable to sit in a car seat
- Let a pet dog into the car and saw that it could go where it needed
- Saw that vehicle-pushing monsters aren't acting weird
- Spawned monsters inside of a vehicle to make sure that was working too

#### Additional context

Overall this is a pretty big buff for vehicle storage, but I don't really see that as an issue if it's more realistic this way. Additionally, players are constantly needing more storage space as crafting gets more complex and more junk gets added for them to haul around.

I do absolutely think that when swerving at high speeds or colliding with things, unsecured items should go flying around the inside of the vehicle, possibly breaking, getting flung from the car, or hurting the occupants. I suspect it would be difficult to do this without massive lag but it would be super funny.

While working on this, I realized it may be a good idea to add a json field for furniture and vehicle parts that increases the cost to retrieve items stored there. This would be appropriate for high volume storage solutions which have no compartments or shelving, such as the dumpster. It can hold a lot, but getting stuff out of it would be inconvenient.

image

### CODE REVIEW

File: data/json/recipes/other/parts.json [hunk 1]
```diff
@@ -763,7 +763,7 @@
       { "proficiency": "prof_welding_basic", "skill_penalty": 0.5 },
       { "proficiency": "prof_welding" }
     ],
-    "components": [ [ [ "pipe", 4 ] ], [ [ "spring", 2 ] ], [ [ "cotton_patchwork", 20 ], [ "sheet", 1 ] ], [ [ "pipe_fittings", 4 ] ] ]
+    "components": [ [ [ "pipe", 4 ] ], [ [ "spring", 2 ] ], [ [ "nylon", 20 ], [ "sheet_nylon", 1 ] ], [ [ "pipe_fittings", 4 ] ] ]
```
Comment (Karol1223) [inline data/json/recipes/other/parts.json:766]:
```suggestion
    "components": [ [ [ "pipe", 4 ] ], [ [ "spring", 2 ] ], [ [ "nylon", 20 ], [ "sheet_nylon", 3 ] ], [ [ "pipe_fittings", 4 ] ] ]
```
``sheet`` isn't a sheet of cotton, it's literally like a bedsheet. The equivalent of 20 nylon patches would be 2.5 nylon sheets

File: data/json/recipes/other/parts.json [hunk 2]
```diff
@@ -790,7 +790,55 @@
...
+    "reversible": { "time": "20 m" },
+    "autolearn": true,
+    "//": "20cm weld",
+    "using": [ [ "sewing_standard", 50 ], [ "welding_standard", 20 ] ],
+    "proficiencies": [
+      { "proficiency": "prof_metalworking" },
+      { "proficiency": "prof_welding_basic", "skill_penalty": 0.5 },
+      { "proficiency": "prof_welding" }
+    ],
+    "components": [ [ [ "pipe", 4 ] ], [ [ "spring", 2 ] ], [ [ "nylon", 20 ], [ "sheet_nylon", 1 ] ], [ [ "pipe_fittings", 4 ] ] ]
```
Comment (Karol1223) [inline data/json/recipes/other/parts.json:814]:
```suggestion
    "components": [ [ [ "pipe", 4 ] ], [ [ "spring", 2 ] ], [ [ "nylon", 20 ], [ "sheet_nylon", 3 ] ], [ [ "pipe_fittings", 4 ] ] ]
```

**Review by Karol1223 — CHANGES_REQUESTED** [review]

**Comment by PatrikLundell** [comment]
I still don't think seats should both act as large storages at the same time as a passenger seat (or bed). While you could presumably (no personal experience, obviously) strap in a corpse in a passenger seat, trying to stack the same volume of random stuff on the seat will be quite dangerous, as stuff would topple off the tile during driving and some of it could end up under the driver's pedals.

**Comment by fairyarmadillo** [comment]
> I still don't think seats should both act as large storages at the same time as a passenger seat (or bed). While you could presumably (no personal experience, obviously) strap in a corpse in a passenger seat, trying to stack the same volume of random stuff on the seat will be quite dangerous, as stuff would topple off the tile during driving and some of it could end up under the driver's pedals.

70 liters isn't large storage. I've driven with groceries in my lap and I have driven with unsecured furniture and huge piles of groceries, boxes of books, and God knows what else in the passenger seat. There were no problems, and it's hardly less safe than anything else characters do.

**Comment by fairyarmadillo** [comment]
And to be clear, this wasn't just down the street and back. I moved over a thousand miles with all of my earthly belongings loaded up this way at high speeds on the freeway.

**Comment by OldFriendEr** [comment]
putting a corpse or tons of junk on the seat is ok in my opinion, but what about storing 60L of items on the seat and then sitting on it and taking controls? 
I think the main issue with seat item capacity is that the items and living creatures do not share the space on the seat.

**Comment by fairyarmadillo** [comment]
> putting a corpse or tons of junk on the seat is ok in my opinion, but what about storing 60L of items on the seat and then sitting on it and taking controls? I think the main issue with seat item capacity is that the items and living creatures do not share the space on the seat.

I addressed this in the PR.

**Comment by fairyarmadillo** [comment]
Working on a solution to characters not counting as cargo. I think I've almost got it, will reopen when I do.

**Comment by fairyarmadillo** [comment]
I thought I had a very clever solution that worked quite well, and then I discovered that the stuff I edited in game.cpp was at the avatar level and not character. I have no idea how NPCs move.

**Comment by fairyarmadillo** [comment]
I now know way more about NPCs than I ever cared to.

**Comment by fairyarmadillo** [comment]
There we go. All tests are passing except a mapgen test which is failing for all PRs and has nothing to do with this.

**Comment by fairyarmadillo** [comment]
> > The vehicle part is described as one square meter in size
> 
> This description is clearly out of line with other handling of vehicle tile size, e.g. #68603.

Tiles are not consistently sized and never will be, it's all just ballparked between one and three square meters. This means in cases where I don't have a real world item (such as a wheelbarrow or specific vehicle trunk) to go off of, I am forced to make somewhat arbitrary decisions.

If you start trying to do otherwise, you realize that all of our cars are like twice as big as they should be, tile-wise, but if they weren't as big as they are, we wouldn't be able to fit the necessary number of characters and seats inside, because in this game a human takes up an entire tile, just like a rat or an elephant. It's unfortunately necessarily nonsensical.

An aisle usially means something like the aisle of a city bus, which is about that wide in my city. Since that's the case they're most often placed in the world as, that's what I have gone with. It also matched the existing description. It makes logical sense that such a space could hold as much as it does, so that's what we have.

**Comment by ehughsbaird** [comment]
As far as I can tell, the established standard is vehicle parts are 0.5x0.5m square. For things like storage volume especially, this seems like an appropriate measure to use.

**Comment by fairyarmadillo** [comment]
There isn't really any kind of standard like that. A vehicle bed for instance would have to be almost two meters long and easily a meter wide, but it only occupies one tile. A wing mirror is a few inches across, still fills a tile.

**Comment by ehughsbaird** [comment]
I have reason to believe there is: https://github.com/CleverRaven/Cataclysm-DDA/pull/68705#issuecomment-1765733368

**Comment by fairyarmadillo** [comment]
🤷‍♀️ Even so it doesn't affect anything other than a single comment line in the json that I didn't write. I wildly undershot the cargo size because 333 liters seemed excessive and it looked like I was correct to do so. 

A trunk, if it's a .5 meter space, holds 300 liters. It is able to do this because it's enclosed. An aisle, if it's a .5 meter space, holds 100 liters. It holds less than a trunk because it's not enclosed and items piled high enough would spill out into adjacent tiles. Makes sense to me.

I don't think changing a single number in a comment line that players can't even see is worth opening a new PR for.

**Comment by ehughsbaird** [comment]
Sorry, I was just trying to notify that what that aisle is saying is (by a statdard partially applied to some parts) wrong. I wasn't trying to suggest changes or anything (it's a comment on a closed PR, I wouldn't be surprised if nobody read it), unless you thought changes were necessary as a result.

**Comment by fairyarmadillo** [comment]
I changed the comment in #70393

**Comment by TinglePan** [comment]
I dragged my travois into a deadend and found myself stuck, with prompt "There's not enough room for you to fit there". Kinda annoying.
Me be like:
###
#@#
#o#
"@" is me, "o"is my travois.

**Comment by fairyarmadillo** [comment]
> I dragged my travois into a deadend and found myself stuck, with prompt "There's not enough room for you to fit there". Kinda annoying. Me be like: "###" "#@#" "#o#" "@" is me, "o"is my travois.

Thank you for bringing this to my attention, I'm doing a PR to fix something else tonight and I can make travois work properly too. It's just a couple of sticks on the ground, you should be able to step over it even if it's full.

**Comment by kasanryukin** [comment]
https://safesandmore.com/products/truckvault-floorvault-base-line-for-suv-vehicle-floor-storage-2-compartments?variant=43490782871712

Floor trunks or marketed as floor vaults if you want to do some research into that for a future PR.

**Comment by fairyarmadillo** [comment]
> I dragged my travois into a deadend and found myself stuck, with prompt "There's not enough room for you to fit there". Kinda annoying. Me be like: "###" "#@#" "#o#" "@" is me, "o"is my travois.

This issue will be solved if #70393 is merged.

File: data/json/vehicleparts/vehicle_parts.json [hunk 3]
```diff
@@ -202,7 +213,7 @@
     "damage_modifier": 60,
     "durability": 95,
     "description": "A small but comfortable bed.",
-    "size": "50 L",
+    "size": "100  L",
```
Comment (Procyonae) [inline data/json/vehicleparts/vehicle_parts.json:216]: Surprised this passed tests you appear to have added an extra space after the number (More interested in tests catching this in future than a change to this specifically)

Comment (fairyarmadillo) [reply data/json/vehicleparts/vehicle_parts.json:216]: Huh, yeah. I have fixed it in #70560 but that's something that shouldn't be getting through.

**Comment by maniak1349** [comment]
> * Upgraded floor trunks to 100 liters of storage for parity with aisles.
It seem to still be 87.5L. Did you forgot to change it?
Also right now floor trunk makes no sense whatsoever - it requires the same amount of time-effort-additional-resource to install, have the same durability and walk speed, but requires specific part instead of readily available sheet metal, weights more and accepts less cargo (same cargo when fixed).

**Comment by kasanryukin** [comment]
The floor trunk is meant as a walkable option when it's full, unlike an
aisle which blocks passage when full.

**Comment by maniak1349** [comment]
> he floor trunk is meant as a walkable option when it's full, unlike an aisle which blocks passage when full.

Oh, I see. Didn't know that. That's a fair game then.
Also is there a way to tell a tileset to show full parts like that differently?

**Comment by kasanryukin** [comment]
Not presently.

**Pull request merged**